src/soc/ibm/power9/fsi.c

```diff
@@ -0,0 +1,317 @@
+
+	/* FSI Control Registers */
+	FSI_MMODE_000  = 0x000,
+	FSI_MLEVP0_018 = 0x018,
+	FSI_MRESP0_0D0 = 0x0D0,
+	FSI_MESRB0_1D0 = 0x1D0,
+	FSI_MECTRL_2E0 = 0x2E0,
+
+	/* MFSI Ports (512KB for each of 8 slaves) */
+	MFSI_PORT_0 = 0x080000,
+
+	/* FSI Slave Register */
+	SLAVE_REGS = 0x000800,
+	SMODE_00   = SLAVE_REGS | 0x00,
+	SLRES_34   = SLAVE_REGS | 0x34,
+
+	/* Bitmasks for OPB status register */
+	OPB_STAT_BUSY       = 0x0001000000000000, // 15 is the Busy bit
+	OPB_STAT_READ_VALID = 0x0002000000000000, // 14 is the Valid Read bit
+	OPB_STAT_ERRACK     = 0x0010000000000000, // 11 is OPB errAck
+	OPB_STAT_ANYERR     = 0x8000000000000000, // 0 is Any error
+	OPB_STAT_ERR_OPB    = 0x7FEC000000000000, // 1:10,12:13 are OPB errors
+	OPB_STAT_ERR_CMFSI  = 0x0000FC0000000000, // 16:21 are cMFSI errors
+	OPB_STAT_ERR_MFSI   = 0x000000FC00000000, // 24:29 are MFSI errors
+
+	OPB_STAT_NON_MFSI_ERR = (OPB_STAT_ERR_OPB |
+				 OPB_STAT_ERRACK |
+				 OPB_STAT_ANYERR),
+	OPB_STAT_ERR_ANY      = (OPB_STAT_NON_MFSI_ERR |
+				 OPB_STAT_ERR_CMFSI |
+				 OPB_STAT_ERR_MFSI),
+};
+
+/* Gets base SCOM address for FSI communication */
+static uint64_t get_opb_offset(uint8_t chip)
+{
+	/* cMFSI 0 and MFSI */
+	enum { FSI2OPB_OFFSET_0 = 0x00020000 };
+
+	const uint8_t port = (chip == 1 ? 1 : 0);
+	return FSI2OPB_OFFSET_0 + port * 0x10;
+}
+
+static void reset_pib2opb(void)
+{
+	const uint64_t opb_offset = get_opb_offset(/*chip=*/0);
+
+	write_scom(opb_offset | OPB_REG_RES, 0x8000000000000000);
+	write_scom(opb_offset | OPB_REG_STAT, 0x8000000000000000);
+}
+
+static void reset_port(uint8_t port)
+{
+	enum {
+		/* FSI2PIB Engine (SCOM) */
+		FSI2PIB_ENGINE   = 0x001000,
+		FSI2PIB_RESET    = FSI2PIB_ENGINE | 0x18, // see CFAM 1006
+		FSI2PIB_COMPMASK = FSI2PIB_ENGINE | 0x30, // see CFAM 100C
+		FSI2PIB_TRUEMASK = FSI2PIB_ENGINE | 0x34, // see CFAM 100D
+	};
+
+	const uint8_t chip = 0;
+
+	uint32_t compmask;
+	uint32_t truemask;
+
+	uint32_t master_ctrl_reg = MFSI_CONTROL_REG;
+
+	/* Slave offset is determined by which port it is on */
+	uint64_t slave_offset = MFSI_PORT_0 * (port + 1);
+
+	/*
+	 * Reset the bridge to clear up the residual errors
+	 * 0=Bridge: General reset
+	 */
+	write_fsi(chip, master_ctrl_reg | FSI_MESRB0_1D0, 0x80000000);
+
+	/* Perform error reset on Centaur fsi slave: write 0x4000000 to addr=834 */
+	write_fsi(chip, SLRES_34, 0x4000000);
+
+	/* Need to save/restore the true/comp masks or the FSP will get annoyed */
+	compmask = read_fsi(chip, slave_offset | FSI2PIB_COMPMASK);
+	truemask = read_fsi(chip, slave_offset | FSI2PIB_TRUEMASK);
+
+	/* Then, write arbitrary data to 1018 (putcfam 1006) to reset any
+	 * pending FSI2PIB errors */
+	write_fsi(chip, slave_offset | FSI2PIB_RESET, 0xFFFFFFFF);
+
+	/* Restore the true/comp masks */
+	write_fsi(chip, slave_offset | FSI2PIB_COMPMASK, compmask);
+	write_fsi(chip, slave_offset | FSI2PIB_TRUEMASK, truemask);
+}
+
+static void init_fsi_port(uint8_t port)
+{
+	enum { FSI_MSENP0_018 = 0x018 };
```

Comment:
Rest of `FSI_*` registers is listed at the top of this file, maybe move it there for consistency.